sdk/go/pulumi/resource.go

```diff
@@ -369,36 +431,95 @@ func Providers(o ...ProviderResource) ResourceOption {
 
 // Timeouts is an optional configuration block used for CRUD operations
 func Timeouts(o *CustomTimeouts) ResourceOption {
-	return resourceOption(func(ro *resourceOptions) {
+	return resourceOption(func(ctx context.Context, ro *resourceOptions) error {
 		ro.CustomTimeouts = o
+		return nil
 	})
 }
 
 // Transformations is an optional list of transformations to be applied to the resource.
 func Transformations(o []ResourceTransformation) ResourceOption {
-	return resourceOption(func(ro *resourceOptions) {
+	return resourceOption(func(ctx context.Context, ro *resourceOptions) error {
 		ro.Transformations = append(ro.Transformations, o...)
+		return nil
 	})
 }
 
 // URN_ is an optional URN of a previously-registered resource of this type to read from the engine.
 //nolint: golint
 func URN_(o string) ResourceOption {
-	return resourceOption(func(ro *resourceOptions) {
+	return resourceOption(func(ctx context.Context, ro *resourceOptions) error {
 		ro.URN = o
+		return nil
 	})
 }
 
 // Version is an optional version, corresponding to the version of the provider plugin that should be used when
 // operating on this resource. This version overrides the version information inferred from the current package and
 // should rarely be used.
 func Version(o string) ResourceOrInvokeOption {
-	return resourceOrInvokeOption(func(ro *resourceOptions, io *invokeOptions) {
+	return resourceOrInvokeOption(func(ctx context.Context, ro *resourceOptions, io *invokeOptions) error {
 		switch {
 		case ro != nil:
 			ro.Version = o
 		case io != nil:
 			io.Version = o
 		}
+		return nil
 	})
 }
+
+type ResourceInput interface {
+	Input
+
+	ToResourceOutput() ResourceOutput
+	ToResourceOutputWithContext(context.Context) ResourceOutput
+}
+
+type plainResourceInput struct {
+	resource Resource
+}
+
+func (pri *plainResourceInput) ElementType() reflect.Type {
+	return resourceType
+}
+
+func (pri *plainResourceInput) ToResourceOutput() ResourceOutput {
+	var ctx Context
+	out, ok, _ := ctx.NewOutput()
+	ok(pri.resource)
+	return ResourceOutput{out.getState()}
+}
+
+func (pri *plainResourceInput) ToResourceOutputWithContext(ctx context.Context) ResourceOutput {
+	return pri.ToResourceOutput()
+}
+
+// TODO should we just include ResourceInput in Resource and deal with any breakage?
+func NewResourceInput(resource Resource) (ResourceInput, error) {
+	if resource == nil {
+		return nil, fmt.Errorf("NewResourceInput cannot be called with a nil resource")
+	}
+	return &plainResourceInput{resource}, nil
+}
+
+// TODO - helps incremental refactoring, but can we completely remove
+// this function?
+func awaitResourceInputMAGIC(ctx context.Context, ri ResourceInput) (Resource, []Resource, error) {
+	result, known, _, deps, err := ri.ToResourceOutput().await(ctx)
+
+	if !known {
```

Comment:
Should we in fact support unknowns instead of failing, for example skipping unknown resources from lists, or making unknown parent as "no parent" (which can be tricky as preview/update will re-parent, change URN)